### ISSUE TITLE
Feature: pLM-BLAST in scikit-bio

### ISSUE THREAD

**Issue by valentynbez** [scikit-bio/scikit-bio#2039]
[Class `ProteinEmbedding`](https://github.com/scikit-bio/scikit-bio/blob/545c9859635a5b8514bd9fd599475717c533f108/skbio/embedding/_protein.py#L29-L121) implements a data structure for the storage of protein embeddings. After it, the alignment of embeddings could be implemented for sequence analysis. 

- [pLM-BLAST](https://academic.oup.com/bioinformatics/article/39/10/btad579/7277200?login=true#420964641)
The substitution matrix is constructed from residue-normalized embeddings for both proteins. This substitution matrix can be further used as an input for Needleman-Wunsch alignment or further transformed into a scoring matrix for Smith-Waterman local alignments, providing both local and global alignments in the output. 

The implementation should also be compatible with other parts of scikit-bio. For example, the resulting substation matrices should be compatible with [skbio.sequence.SubstitutionMatrix](https://scikit.bio/docs/latest/generated/skbio.sequence.SubstitutionMatrix.html#skbio.sequence.SubstitutionMatrix). Multiple local alignments from pLM-BLAST might be returned as `TabularMSA` object.

**Comment by valentynbez** [comment]
@mortonjt yes, I'm thinking about `SubstitutionMatrix.from_embedding(emb1, emb2)`. 

For now, alignment algorithms use dicts based on protein alphabet as substitution matrix. Here, we don't exactly have an opportunity to create a `dict` from the resulting substitution matrix, because it does not rely on residues. 
I don't know if a rework of the aligner is possible, the code looks intricate to me.

**Comment by Argusmocny** [comment]
Hi, 

* here is my fork for this issue: https://github.com/Argusmocny/scikit-bio/tree/plmblast 
* for the substitution-matrix the whole is actually 3 vectorized operations in numpy: normalize each embedding row and then matrix multiplication of `emb1 x emb2.T` . The code is here: https://github.com/labstructbioinf/pLM-BLAST/blob/438d45a1ac5fbd7a6363ea1c6b26bbb8b5f8a66e/alntools/numeric.py#L366 - the function is much longer because some array operations are not implemented in `numba`. 
* Maybe I can write another `SubstitutionMatrix` class/template which will work when substitution scores will be contextual eg. depending on position in sequence.

**Comment by qiyunzhu** [comment]
Hello @mortonjt @valentynbez @Argusmocny  Sorry for my late involvement in this valuable conversation.

1. The plan sounds very cool to me! I mean, all of it. Alignments with positional scoring. Substitution matrix from embedding. etc.
2. I agree that adding Numba support requires careful thinking. I personally had good experience with Numba. I want to note the situations where Numba will actually accelerate: slow Python looping. If your code is already vectorized (like the one @Argusmocny pointed to), Numba will make little difference. This is my guess, but @Argusmocny can benchmark it.
3. I have long been planning to implement the global and local pairwise alignment algorithms from scratch in scikit-bio. See my issue #1864 . The prototype code is already there. The plan is moving slowly amid many of my other responsibilities as a faculty member. But if any of you want to give a hand, let's chat!

**Comment by Argusmocny** [comment]
Here is the Substitution matrix and scoring matrix with `SubstitutionMatrix.from_embedding(emb1, emb2)` method https://github.com/Argusmocny/scikit-bio/blob/e39550d433bf5e4b5b939ab31f7a1d1d90134c6e/skbio/embedding/_substitution.py#L16 
As far as I understand some of basic `SubstitutionMatrix` fuctionalities will not be compatible with positional based scores, so i moved it to `skbio/embedding/_substitution.py` for now.  What do you think about this?

**Comment by mortonjt** [comment]
Hi @Argusmocny thank you for putting this together. I think it is a good start.

A couple of thoughts here
1. `SubstitutionMatrix` here would have name conflict with the `skbio.sequence.substitinon_matrix`. Maybe we could call it `PositionSpecificSubstitutionMatrix` or `EmbeddingSubstitionMatrix`
2. Do you have unit tests that confirm that the alignment capabilities work with the proposed substation matrix functionality?
3. It may not be a bad idea to look into having the proposed `SubstitionMatrix` be a subclass of the [original class](https://github.com/Argusmocny/scikit-bio/blob/e39550d433bf5e4b5b939ab31f7a1d1d90134c6e/skbio/sequence/_substitution.py#L16) -- or if that is out of scope, port over the relevant attributes so that the APIs are mirrored.

The only concern that I have is that folks may have trouble finding the proposed `SubstitionMatrix` functionality if it is tucked away in the `skbio.embedding` section. It may not be a bad idea to update the toctree / documentation in the alignment section to make this link clear.
